### PR TITLE
Add new refactoring feature to wrap chains of fluent calls.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Looks like this:

image

image

image

image

Todo:

- [x] Tests.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @sharwell @kendrahavens @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
@sharwell This is ready for review.  Of particular interest to me is a discussion of the overall wrapping algorithm and if it gives behavior that we think is desirable.  I tried a few approaches and ended up with the current one that i like.  Core benefits are primarily that it will work on effectively any dotted call chain, and it will align the `.M(...)` parts uniformly.  

I've tried to explain how it works in the comments.

File: src/EditorFeatures/CSharp/Wrapping/BinaryExpression/CSharpBinaryExpressionWrapper.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ public CSharpBinaryExpressionWrapper()
         {
         }
 
-        protected override SyntaxTriviaList GetNewLineBeforeOperatorTrivia(SyntaxTriviaList newLine)
+        public override SyntaxTriviaList GetNewLineBeforeOperatorTrivia(SyntaxTriviaList newLine)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/Wrapping/BinaryExpression/CSharpBinaryExpressionWrapper.cs]: required because i moved the abstract method to a locatoin that isn't accessible through protected access.

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 2]
```diff
@@ -1598,4 +1598,13 @@ This version used in: {2}</value>
...
     <value>Use expression body for local functions</value>
   </data>
+  <data name="Unwrap calls" xml:space="preserve">
+    <value>Unwrap calls</value>
+  </data>
+  <data name="Wrap_calls" xml:space="preserve">
+    <value>Wrap calls</value>
+  </data>
+  <data name="Wrap_long_calls" xml:space="preserve">
+    <value>Wrap long calls</value>
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/FeaturesResources.resx]: Alternative names:

1. Unwrap call chain
2. Wrap call chain
3. Wrap long call chain

Comment (alrz) [reply src/Features/Core/Portable/FeaturesResources.resx]: definitely `Wrap call chain`

long lines doesn't communicate that this is operating on, well, call chains.

Comment (alrz) [reply src/Features/Core/Portable/FeaturesResources.resx]: I didn't noticed there is two options, how about

- Wrap call chains
- Wrap call chains on separate lines

"long" there was a little confusing to me.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/FeaturesResources.resx]: both of those would be wrapping to separate lines.  So it's unclear to me how those two actually differ.

We use the 'long' term already for the existing wrapping functions.  Specifically "wrap long parameter list" "wrap long argument list" and "wrap long expression".  It basically means "wrap once you go past 120 chars".

**Comment by sharwell** [comment]
> ... and it will align …

It should indent the continuation by one indent. It should not make an attempt to align if that is not the offset of the first `.`.

**Comment by CyrusNajmabadi** [comment]
> It should indent the continuation(s) by one indent. It should not make an attempt to align if that is not the offset of the first ..

Why?

Alighnment is super common.  You can see thousands of examples of it in the IDE and compiler codebase today.

If this was about just indenting, then you wouldn't really need the feature.  You just hit enter.  The value here is in being able to make a change and have an easy way to get the entire chained-call to be consistently formatted.

**Comment by JoeRobich** [comment]
> Note: i'm amenable to having both be offered (just like we do with arg/param-lists). But the alignment case (like with params/args) is the really helpful one.

@CyrusNajmabadi I think we would probably want both. Personally I prefer the wrap and indent over wrap and align, but I understand why they are both valuable.

**Comment by CyrusNajmabadi** [comment]
Note: the more i think aboutthis, the more this seems very similar to 'wrapping parameters/arguments'.  i.e. there's merit behind lots of these different ways to style things.  Including all of the following:

1. Wrap all but first, and align. i.e.

```c#
var foo = expr.Whatever()
              .ContinueWith()
              .Etc();
```

2. Wrap all but first, and indent. i.e.

```c#
var foo = expr.Whatever()
    .ContinueWith()
    .Etc();
```

3. Wrap all:

```c#
var foo = expr
    .Whatever()
    .ContinueWith()
    .Etc();
```

I've looked and all these forms are represented in the Roslyn codebase itself in significant quantities.  Enough that it's clear there is no preferred style, and that people use whichever they feel is appropriate depending on context.

@sharwell @JoeRobich would you have issues with this wrapping feature jsut behaving the same was as param/arg wrapping in this regard?

**Comment by JoeRobich** [comment]
> @sharwell @JoeRobich would you have issues with this wrapping feature jsut behaving the same was as param/arg wrapping in this regard?

@CyrusNajmabadi I think that would be a great idea.

File: src/EditorFeatures/CSharp/Wrapping/BinaryExpression/CSharpBinaryExpressionWrapper.cs (hunk 1 above)
Comment (Neme12) [reply src/EditorFeatures/CSharp/Wrapping/BinaryExpression/CSharpBinaryExpressionWrapper.cs]: I still maintain that a code fix to expand accessibility would be useful in practice...

**Comment by CyrusNajmabadi** [comment]
closing

**Comment by CyrusNajmabadi** [comment]
Opening.

**Comment by CyrusNajmabadi** [comment]
@sharwell Can we move forward on this?

**Comment by CyrusNajmabadi** [comment]
@sharwell @jinujoseph What's a good way to get traction on this?

**Comment by sharwell** [comment]
Design review conclusion: if possible, we'd like this refactoring to follow the same pattern that argument wrapping does, i.e. show both indentation options and allow the user to pick. If it's not practical to implement that change from the current state, we'd still be in favor of moving this pull request forward in its current form, even though some users would be uncomfortable making use of it.

**Comment by sharwell** [comment]
@CyrusNajmabadi can you give a brief overview of the work required to implement the change described above, so we can determine the best time to complete it?

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi can you give a brief overview of the work required to implement the change described above, so we can determine the best time to complete it?

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxKindsService.cs [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
 
 namespace Microsoft.CodeAnalysis.CSharp.LanguageServices
 {
-    [ExportLanguageService(typeof(ISyntaxKindsService), LanguageNames.CSharp), Shared]
-    internal sealed class CSharpSyntaxKindsService : ISyntaxKindsService
+    internal sealed class CSharpSyntaxKindsService : AbstractSyntaxKindsService
     {
+        public static readonly CSharpSyntaxKindsService Instance = new CSharpSyntaxKindsService();
+
         [ImportingConstructor]
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxKindsService.cs]:
```suggestion
```

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxKindsService.cs [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
         public CSharpSyntaxKindsService()
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxKindsService.cs]:
```suggestion
         private CSharpSyntaxKindsService()
```

**Comment by CyrusNajmabadi** [comment]
@jinujoseph @JoeRobich @sharwell This is ready for review.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxKindsService.cs [hunk 5]
```diff
@@ -18,4 +25,15 @@ internal interface ISyntaxKindsService : ILanguageService
         /// </summary>
         int LogicalOrExpression { get; }
     }
+
+    internal abstract class AbstractSyntaxKindsService : ISyntaxKindsService
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxKindsService.cs]: ❔ Why is this needed? Seems like the interface would be sufficient.

File: src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxKindsService.vb [hunk 6]
```diff
@@ -1,20 +1,24 @@
...
 
 Namespace Microsoft.CodeAnalysis.VisualBasic.LanguageServices
-    <ExportLanguageService(GetType(ISyntaxKindsService), LanguageNames.VisualBasic), [Shared]>
     Friend NotInheritable Class VisualBasicSyntaxKindsService
-        Implements ISyntaxKindsService
+        Inherits AbstractSyntaxKindsService
+
+        Public Shared ReadOnly Instance As New VisualBasicSyntaxKindsService()
 
         <ImportingConstructor>
```
Comment (sharwell) [inline src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxKindsService.vb]:
```suggestion
```

File: src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxKindsService.vb [hunk 7]
```diff
@@ -1,20 +1,24 @@
...
         Public Sub New()
```
Comment (sharwell) [inline src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxKindsService.vb]:
```suggestion
         Private Sub New()
```

File: src/EditorFeatures/CSharpTest/Wrapping/ChainedExpressionWrappingTests.cs [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+    void Bar() {
+        [||]the.quick();
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsWrapping)]
+        public async Task TestWithEnoughChunks()
+        {
+            await TestAllWrappingCasesAsync(
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Wrapping/ChainedExpressionWrappingTests.cs]: ❔ Should this also include the wrap-without-alignment case?

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+            }
+
+            // We're at the top of something that looks like it could be part of a chained
+            // expression.  Break it into the individual chunks.  We need to have at least
+            // two chunks or this to be worth wrapping.
+            //
+            // i.e. if we only have `this.Goo(...)` there's nothing to wrap.  However, we can
+            // wrap when we have `this.Goo(...).Bar(...)`.
+            var chunks = GetChainChunks(node);
+            if (chunks.Length <= 1)
```
Comment (sharwell) [inline src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]:
```suggestion
            if (chunks.Length < 2)
```

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+
+                // Had a chunk after this one.  Record the current chunk, move to the start
+                // of the next one, and then keep going.
+                chunks.Add(GetSubRange(pieces, currentChunkStart, end: nextChunkStart));
+                currentChunkStart = nextChunkStart;
+            }
+        }
+
+        /// <summary>
+        /// Looks for the next sequence of `. Name (ArgList)`.  Note, except for the first
```
Comment (sharwell) [inline src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]: ❕ `` ` `` is not a supported formatting character in XML documentation comments. Wrap code with `<c>`...`</c>` instead.

```suggestion
        /// Looks for the next sequence of <c>. Name (ArgList)</c>.  Note, except for the first
```

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.Editor.Wrapping.ChainedExpression
+{
+    /// <summary>
+    /// Finds and wraps 'chained' expressions.  For the purpose of this feature, a chained
+    /// expression is built out of 'chunks' where each chunk is of the form
+    ///
+    /// <c>
```
Comment (sharwell) [inline src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]: 📝 `<c>` is the inline code element. If you mean for these to be separate blocks treated similar to paragraphs, use `<code>`.

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+    /// 
+    /// <c>
+    ///     . name1 (arglist1) remainder1
+    ///     . name2 (arglist2) remainder2
+    /// </c>
+    /// 
+    /// Note: for the sake of simplicity, (arglist) is used both for the argument list of
+    /// an InvocationExpression and an ElementAccessExpression.
+    /// 
+    /// 'remainder' is all the postfix expression that can follow `. name (arglist)`.  i.e.
```
Comment (sharwell) [inline src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]: 📝 Text paragraphs should be wrapped in `<para>` if there is more than one.

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+    /// member-access expressions, conditional-access expressions, etc.  Effectively, anything
+    /// the language allows at this point as long as it doesn't start another 'chunk' itself.
+    /// 
+    /// This approach gives an intuitive wrapping algorithm that matches the common way
+    /// many wrap dotted invocations, while also effectively not limiting the wrapper to
+    /// only simple forms like `.a(...).b(...).c(...)`.
```
Comment (sharwell) [inline src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]:
```suggestion
    /// only simple forms like <c>.a(...).b(...).c(...)</c>.  
```

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+                // Had a chunk after this one.  Record the current chunk, move to the start
+                // of the next one, and then keep going.
+                chunks.Add(GetSubRange(pieces, currentChunkStart, end: nextChunkStart));
+                currentChunkStart = nextChunkStart;
+            }
+        }
+
+        /// <summary>
+        /// Looks for the next sequence of `. Name (ArgList)`.  Note, except for the first
+        /// chunk, this cannot be of the form `? . Name (ArgList)` as we do not want to
```
Comment (sharwell) [inline src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]:
```suggestion
        /// chunk, this cannot be of the form <c>? . Name (ArgList)</c> as we do not want to 
```

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+                // of the next one, and then keep going.
+                chunks.Add(GetSubRange(pieces, currentChunkStart, end: nextChunkStart));
+                currentChunkStart = nextChunkStart;
+            }
+        }
+
+        /// <summary>
+        /// Looks for the next sequence of `. Name (ArgList)`.  Note, except for the first
+        /// chunk, this cannot be of the form `? . Name (ArgList)` as we do not want to 
+        /// wrap before a dot in a `?.` form.  This doesn't matter for the first chunk as
```
Comment (sharwell) [inline src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]:
```suggestion
        /// wrap before a dot in a <c>?.</c> form.  This doesn't matter for the first chunk as
```

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Recursively walks down 'node' decomposing it into the individual tokens and
```
Comment (sharwell) [inline src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]:
```suggestion
        /// Recursively walks down <paramref name="node"/> decomposing it into the individual tokens and 
```

File: src/EditorFeatures/VisualBasic/Wrapping/ChainedExpression/VisualBasicChainedExpressionWrapper.vb [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.Wrapping.ChainedExpression
+    Friend Class VisualBasicChainedExpressionWrapper
+        Inherits AbstractChainedExpressionWrapper(Of NameSyntax, ArgumentListSyntax)
+
+        Public Sub New()
+            MyBase.New(VisualBasicIndentationService.WithoutParameterAlignmentInstance, VisualBasicSyntaxFactsService.Instance)
+        End Sub
+
+        Protected Overrides Function GetNewLineBeforeOperatorTrivia(newLine As SyntaxTriviaList) As SyntaxTriviaList
+            Return newLine.InsertRange(0, {SyntaxFactory.WhitespaceTrivia(" "), SyntaxFactory.LineContinuationTrivia("_")})
```
Comment (sharwell) [inline src/EditorFeatures/VisualBasic/Wrapping/ChainedExpression/VisualBasicChainedExpressionWrapper.vb]: > SyntaxFactory.WhitespaceTrivia(" ")

❔  Is this equivalent to `SyntaxFactory.ElasticSpace`?

File: src/EditorFeatures/VisualBasicTest/Wrapping/ChainedExpressionWrappingTests.vb [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+"class C
+    sub Bar()
+        [||]the.quick()
+    end sub
+end class")
+        End Function
+
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsWrapping)>
+        Public Async Function TestWithEnoughChunks() As Task
+            Await TestAllWrappingCasesAsync(
```
Comment (sharwell) [inline src/EditorFeatures/VisualBasicTest/Wrapping/ChainedExpressionWrappingTests.vb]: 💡 Consider adding cases for wrapping without alignment

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 19]
```diff
@@ -38,6 +39,8 @@ public SyntaxTrivia ElasticMarker
         public SyntaxTrivia ElasticCarriageReturnLineFeed
             => SyntaxFactory.ElasticCarriageReturnLineFeed;
 
+        public ISyntaxKindsService SyntaxKinds { get; } = CSharpSyntaxKindsService.Instance;
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: 💡 Consider making `ISyntaxKindsService` exclusively provided through this service as `ISyntaxKinds`, and eliminating it as a standalone MEF export.

File: src/EditorFeatures/CSharpTest/Wrapping/ChainedExpressionWrappingTests.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Wrapping/ChainedExpressionWrappingTests.cs]: yup.  i'm updating these as we speak :)

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs (hunk 11 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]: thanks!

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]: resolved.

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]: #mybrainnuderstandsthisbetter :)

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]: Fixed everywhere.

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs (hunk 16 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]: Done.

File: src/EditorFeatures/VisualBasicTest/Wrapping/ChainedExpressionWrappingTests.vb (hunk 18 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasicTest/Wrapping/ChainedExpressionWrappingTests.vb]: done. #resolved.

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxKindsService.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxKindsService.cs]: Done!

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxKindsService.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxKindsService.cs]: So this is due to pain in authoring stuff in VB.  In VB you have to have the explicit interface impl form.  This approach allows hte VB impl to just be simple with overrides.  I agree that's marginal.  So i can remove.  But i've found htis much more pleasant from an authoring perspective as things like implement-interface "work", but don't necessarily respect initial order.  So yo have to go manually use it, then move things around.

File: src/EditorFeatures/VisualBasic/Wrapping/ChainedExpression/VisualBasicChainedExpressionWrapper.vb (hunk 17 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasic/Wrapping/ChainedExpression/VisualBasicChainedExpressionWrapper.vb]: Good question.  Tried it and everything went off the rails.  The VB formatter started emitting newlines *before* the elasticspace.  No clue why.  Don't want to spend time figuring out :D

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs (hunk 19 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: I like having it both ways.  So i'm going to keep it hte same for now.

File: src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxKindsService.vb (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxKindsService.vb]: Done!

**Comment by CyrusNajmabadi** [comment]
@sharwell Feedback addressed.

**Comment by CyrusNajmabadi** [comment]
@sharwell This is ready for another pass/merge if you're ready.

File: src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+            {
+                var unformattable = await ContainsUnformattableContentAsync(
+                    document, chunk, cancellationToken).ConfigureAwait(false);
+                if (unformattable)
+                {
+                    return null;
+                }
+            }
+
+            // Looks good.  Crate the action computer which will actually determine
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Wrapping/ChainedExpression/AbstractChainedExpressionWrapper.cs]:
```suggestion
            // Looks good.  Create the action computer which will actually determine
```

**Comment by CyrusNajmabadi** [comment]
@JoeRobich Yaay!

**Comment by JoeRobich** [comment]
Thanks @CyrusNajmabadi!

**Comment by CyrusNajmabadi** [comment]
Thank you!
